File: Leetcode/Linked List/Two Pointer Technique/removeNthFromEnd.cpp
// https://leetcode.com/explore/learn/card/linked-list/214/two-pointer-technique/1296/

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        if(n <= 0 ){
            return head;
        }
        auto slow = &head;
        ListNode* fast = head;
        while(n--){
            fast = fast -> next;
        }
        while(fast){
            slow = &(*slow) ->next;
            fast = fast -> next;
        }
        fast = *slow;
        *slow = (*slow)->next;
        delete fast;   
        return head;
    }
};

File: Leetcode/Problems/Medium/removeNthFromEnd.py
# https://leetcode.com/problems/remove-nth-node-from-end-of-list/
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        slow = head
        fast = head
        while(n > 0):
            fast = fast.next
            n -= 1
        if not fast:
            return head.next
        while(fast.next):
            fast = fast.next
            slow = slow.next
        slow.next = slow.next.next
        return head
        
